src/passes/Asyncify.cpp

```diff
@@ -515,52 +512,47 @@ class ModuleAnalyzer {
         }
       });
 
-    map.swap(scanner.map);
-
     // Functions in the blacklist are assumed to not change the state.
-    for (auto& name : blacklist.names) {
-      if (auto* func = module.getFunctionOrNull(name)) {
-        map[func].canChangeState = false;
+    for (auto& pair : scanner.map) {
+      auto* func = pair.first;
+      auto& info = pair.second;
+      if (blacklist.match(func->name)) {
+        info.inBlacklist = true;
+        info.canChangeState = false;
       }
     }
 
-    // Remove the asyncify imports, if any.
-    for (auto& pair : map) {
+    // Remove the asyncify imports, if any, and any calls to them.
+    std::vector<Name> toDelete;
+    for (auto& pair : scanner.map) {
       auto* func = pair.first;
+      auto& callsTo = pair.second.callsTo;
       if (func->imported() && func->module == ASYNCIFY) {
-        module.removeFunction(func->name);
+        toDelete.push_back(func->name);
       }
-    }
-
-    // Find what is called by what.
-    for (auto& pair : map) {
-      auto* func = pair.first;
-      auto& info = pair.second;
-      for (auto* target : info.callsTo) {
-        map[target].calledBy.insert(func);
+      std::vector<Function*> toDelete;
+      for (auto* target : callsTo) {
+        if (target->imported() && target->module == ASYNCIFY) {
+          toDelete.push_back(target);
```

Comment:
- Aren't these functions also added to `toDelete` in the code above
```cpp
if (func->imported() && func->module == ASYNCIFY) {
  toDelete.push_back(func->name);
}
```
, so each function is added to `toDelete` twice?

- Can't we call `callsTo.erase(target);`, which is called separately below, here as well?

Reply:
Ah, good point, `toDelete` was used twice here, in a confusing way. It was only correct due to scoping... I'll rename them both.

I don't think we can call erase directly in either case, as we may get iterator invalidation. In the first case, `funcsToDelete` after I am renaming it now, we may want to delete foo but foo may be a caller of a function we see later. In the second case, we iterate over `callsTo` so we can't erase in the loop itself (I think? is there some magic c++ feature that makes it safe to do so in some cases?).